_source/_posts/2022-06-16-oauth-java.adoc

```diff
@@ -0,0 +1,389 @@
+  <div class="video">
+    <iframe width="267" height="150" src="https://www.youtube.com/embed/IG2uo4IP1QI" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
+  </div>
+  <div class="video">
+    <iframe width="267" height="150" src="https://www.youtube.com/embed/Bj69rOIwQwY" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
+  </div>
+</div>
+++++
+
+At the end of each video, I show how the https://cli.okta.com[Okta CLI] allows you to create working examples for each framework with the `okta start` command.
+
+If you prefer reading and copy pasta over video, check out these quick tutorials:
+
+- link:/blog/2022/05/23/java-rest-api-quarkus[Kickstart a Java REST API with Quarkus]
+- link:/blog/2022/05/25/java-rest-api-spring-boot[Start a Java REST API with Spring Boot]
+- link:/blog/2022/05/31/java-rest-api-micronaut[Launch a Java REST API with Micronaut]
+- link:/blog/2022/06/02/java-rest-api-helidon[Create a Java REST API with Helidon]
+
+== Java web frameworks without OAuth support
+
+You might get away with saying that JSF and Wicket don't have OAuth support. However, they both build on the Servlet API so it is possible to use them with some of Java's independently-minded security frameworks.
+
+It's possible your favorite framework doesn't have a dedicated security plugin for OAuth, but that's not important if you can get _something_ to work, IMHO.
+
+== Java security frameworks
+
+There are a few Java security frameworks that aren't tied to a specific web framework. I like to call them independently-minded, but honestly, they seem to struggle as volunteer versus funded open source projects.
+
+=== Apache Shiro
+
+Apache Shiro expect to operate in a servlet environment and can work with any framework that's based on servlets and filters. It has many https://shiro.apache.org/integration.html[integrations]&mdash;such as Play, Wicket, and Lift&mdash;but most are dated and unmaintained.
+
+Shiro can be adapted to work with OAuth 2.0 as demonstrated by Brian Demers in link:/blog/2020/05/11/java-shiro-oauth[Build a Secure Java Application with Apache Shiro and OAuth 2.0].
+
+=== Pac4j
+
+Pac4j is a security framework with many sub-projects that implement OAuth 2.0, OIDC, and many others, like SAML.
+
+A gentle reminder:
+
+++++
+{% twitter 1379871695163510795 %}
+++++
+
+When a framework doesn't have OAuth support, I've been able to use Pac4J successfully. For example, with link:/blog/2017/10/31/add-authentication-to-play-framework-with-oidc[Play framework] and link:/blog/2018/09/12/secure-java-ee-rest-api#use-pac4j-to-lock-down-your-java-rest-api[plain ol' Java EE].
+
+You don't hear much about Pac4J in the Java ecosystem. I'm not certain why that is. I've heard it's quality isn't that great and I also believe it suffers because it's not tied to a popular framework.
```

Comment:
```suggestion
You don't hear much about Pac4J in the Java ecosystem. I'm not certain why that is. I've heard its quality isn't that great and I also believe it suffers because it's not tied to a popular framework.
```
its vs it's